Build a regression model.

In [18]:
import numpy as np
import pandas as pd
# first let's get the dataframe 
merged_df=pd.read_csv('C:/Users/Emran Mohammed/Downloads/full_data.csv')
#now we have got a clean data named merged data, let's start building the model
import statsmodels.api as sm
y=merged_df['free_bikes'] # we are predicting number of total bikes in the given station am using 
X=merged_df[['number_of_public_transportation', 'number_of_chinese_restaurants', 'number_of_sport_recreation', 'empty_slots', 'latitude', 'longitude']]
X = sm.add_constant(X) # adding a constant
model = sm.OLS(y,X)
result = model.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.193
Model:                            OLS   Adj. R-squared:                  0.183
Method:                 Least Squares   F-statistic:                     19.60
Date:                Sun, 17 Dec 2023   Prob (F-statistic):           1.47e-20
Time:                        22:40:16   Log-Likelihood:                -1960.7
No. Observations:                 500   AIC:                             3935.
Df Residuals:                     493   BIC:                             3965.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
const     

Provide model output and an interpretation of the results. 

In [19]:
# as we can see that the p value for number_of_sport_recreation is greater than 0.05 which means we will accept the null hypothesis as they have no correlation
# so, we need to construct a new model by removing the column number_of_sport_recreation
y=merged_df['free_bikes'] # we are predicting number of total bikes in the given station am using 
X=merged_df[['number_of_public_transportation', 'number_of_chinese_restaurants', 'empty_slots', 'latitude', 'longitude']]
X = sm.add_constant(X) # adding a constant
model = sm.OLS(y,X)
result = model.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.192
Model:                            OLS   Adj. R-squared:                  0.183
Method:                 Least Squares   F-statistic:                     23.43
Date:                Sun, 17 Dec 2023   Prob (F-statistic):           3.78e-21
Time:                        22:40:57   Log-Likelihood:                -1961.0
No. Observations:                 500   AIC:                             3934.
Df Residuals:                     494   BIC:                             3959.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
const     

In [20]:
# as we can see that the p value for latitude is greater than 0.05 which means we will accept the null hypothesis as they have no correlation
# so, we need to construct a new model by removing the column latitude
y=merged_df['free_bikes'] # we are predicting number of total bikes in the given station am using 
X=merged_df[['number_of_public_transportation', 'number_of_chinese_restaurants', 'empty_slots', 'longitude']]
X = sm.add_constant(X) # adding a constant
model = sm.OLS(y,X)
result = model.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.189
Model:                            OLS   Adj. R-squared:                  0.182
Method:                 Least Squares   F-statistic:                     28.75
Date:                Sun, 17 Dec 2023   Prob (F-statistic):           1.69e-21
Time:                        22:44:06   Log-Likelihood:                -1962.0
No. Observations:                 500   AIC:                             3934.
Df Residuals:                     495   BIC:                             3955.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
const     

In [21]:
# as we can see that the p value for longtiude is greater than 0.05 which means we will accept the null hypothesis as they have no correlation
# so, we need to construct a new model by removing the column longtiude
y=merged_df['free_bikes'] # we are predicting number of total bikes in the given station am using 
X=merged_df[['number_of_public_transportation', 'number_of_chinese_restaurants', 'empty_slots']]
X = sm.add_constant(X) # adding a constant
model = sm.OLS(y,X)
result = model.fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.184
Model:                            OLS   Adj. R-squared:                  0.179
Method:                 Least Squares   F-statistic:                     37.29
Date:                Sun, 17 Dec 2023   Prob (F-statistic):           9.61e-22
Time:                        22:46:20   Log-Likelihood:                -1963.4
No. Observations:                 500   AIC:                             3935.
Df Residuals:                     496   BIC:                             3952.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
const     

now we can say that as all the values for p are less than p so we can reject the null hypothesis and conclude that these variables(POI) have correlation with the column free bikes.
we can interpret the result we got from the above model.
1. The first thing is Adj. R-squared which The model explains about 17.9% of the variance in the number of free bikes (R-squared: 0.184), which is moderate. It suggests that while the included predictors have some influence, there are likely other factors affecting free bike availability.
2. now let's interpret the coefficient of each variables:
     2.1 for number_of_public_transportation (Coefficient: 0.2741): For each additional point of public transportation, the number of free bikes increases by approximately 0.2741. Contrary to initial expectations, the model shows a positive correlation between the number of public transportation options and the number of free bikes. This might suggest that areas with more public transport options could also be areas where bikes are frequently returned, or the total number of bikes in the bike station might be very high
     2.2 for number_of_chinese_restaurants (Coefficient: 0.2554): For each additional Chinese restaurant, the number of free bikes increases by approximately 0.2554. Similar to public transportation, the positive coefficient for the number of Chinese restaurants indicates that stations near more dining options tend to have more free bikes. This could be due to a variety of reasons, such as these areas being popular destinations where bikes are often returned or areas with a different pattern of bike usage.
     2.3 for empty_slots (Coefficient: -0.3277): The number of empty slots is negatively associated with the number of free bikes. For each additional empty slot, the number of free bikes decreases by approximately 0.3277. This relationship aligns more closely with expectations


# Stretch

How can you turn the regression model into a classification model?

In this project, we aim to construct a regression model, where we predicted the continuous value of 'free bikes' at various bike stations, we can transform the problem of predicting the continuous value of 'free bikes' at bike stations (regression) into categorizing stations based on bike availability (classification). my plan to turn the regression model into a classification model is based on ocategorizing bike stations based on bike availability into discrete classes such as 'Low Availability' (less than 5 free bikes), 'Medium Availability' (5 to 15 free bikes), and 'High Availability' (more than 15 free bikes). These categories will be determined based on the data distribution of 'free bikes'. For the classification model, as we did with that of regression model, we will continue to use key features from our regression model, such as number_of_public_transportation, number_of_chinese_restaurants and geographical coordinates. my choice of models is multinomial regression as i am getting more than two outcomes, if it were binary response like only if low or high exist, i will probably use logistic regression. After I am done with constructing model, i will use validation tools inorder to evaluat metrics such as accuracy, precision. This transformation is aimed at providing more actionable insights for operational planning in bike-sharing systems, such as efficient bike redistribution and predicting demand surges. While this approach simplifies the continuous nature of bike availability into categories, it offers a more practical perspective for operational decision-making and strategic planning in urban bike-sharing scenarios.